TLA+ formalization of Percolator.

foreverbell opened the pull request:
Based on the work of @siddontang https://github.com/siddontang/tlaplus-example/blob/master/Percolator.tla.

This TLA+ spec is verified by TLC with these parameters.
* KEY = 1..5
* CLIENT = {"C1", "C2", "C3"}

Also we've proved TypeInvariant with TLAPS.

foreverbell committed:
TLA+ for Percolator.

foreverbell committed:
Add Percolator.cfg to describe TLC parameters.

siddontang commented on Percolator/Percolator.tla:
```diff
@@ -0,0 +1,416 @@
+
+Invariant == TypeInvariant
+
+LEMMA InitStateSatisfiesInvariant ==
```
can we move the proof to the end of the file?

foreverbell replied:
Done.

foreverbell committed:
Move TLAPS proof to bottom.

foreverbell committed:
Merge branch 'master' of https://github.com/pingcap/tla-plus into percolator

foreverbell committed:
Delete out-of-dated Percolator.tla.

siddontang commented on Percolator/Percolator.tla:
```diff
@@ -0,0 +1,417 @@
+  IN
+    \/ /\ hasStaleLock(primary, start_ts)
+       /\ cleanupStaleLock(primary, start_ts)
+    \/ \E k \in secondary :
```
\A here?

foreverbell replied:
Pay attention to the outer "\\/".
This cleans up only one lock, and cleans one by one.. so \E is intended.

siddontang commented on Percolator/Percolator.tla:
```diff
@@ -0,0 +1,417 @@
+\*  2. Or the lock is a secondary lock, and the associated primary key (<= ts)
+\*     has been cleaned up.
+isStaleLock(k, l, ts) ==
+  \/ /\ l.primary = k  \* this is a primary lock.
```
checking ts at first may be better? like
```
/\ l.ts <= ts
/\ \/ primary
   \/ second 
```

siddontang commented on Percolator/Percolator.tla:
```diff
@@ -0,0 +1,417 @@
+         /\ client_state' = [client_state EXCEPT ![c] = "prewriting"]
+         /\ UNCHANGED <<next_ts, key_lock, key_data, key_write, client_ts, client_key>>
+       ELSE
+         /\ cleanup(c)
```
if clean up failed, how can we enter aborted state?

foreverbell replied:
I prefer removing the aborted state for simplicity.

If we cannot clean up any locks, the corresponding client will just stay halted, i.e. it has no following states, stutters in the old state.

siddontang replied:
This is not a full transaction flow, I suggest moving to the aborted state.

foreverbell replied:
As we don't care about liveness property, how about adding one `Abort(c)` action in `Next`? One client can abort at any time if it is not committed, this should complete the transaction workflow, and is what I suggest to do. So I did that.

siddontang replied:
No need to handle aborted in Next like https://github.com/tlaplus/Examples/blob/master/specifications/transaction_commit/TCommit.tla does. 

IMO, it is very strange that a transaction can keep the working state but not go to aborted.

foreverbell replied:
Discussed offline, agreed on lifting `Abort` as a separate action. Added more comments to describe its meaning.

foreverbell committed:
Minor changes.

foreverbell committed:
Minor changes.

foreverbell committed:
Add aborted state.

foreverbell committed:
Minor changes.

foreverbell commented:
I have a follow-up PR to verify the snapshot isolation property. Will send it out once this PR gets merged.

foreverbell committed:
Add more comments for aborted state.

IANTHEREAL commented on Percolator/Percolator.tla:
```diff
@@ -0,0 +1,452 @@
+  \E l \in key_lock[k] : l.ts = ts
+
+\* Returns TRUE if the client encounters no lock, then can do pre-write.
+canPrewrite(c) ==
```
at first glance it make me puzzled, I had to find the definition of `prewrite` to confirm it. 
 it's better to rename other name like `canGoPrewrite` or add some other comments

IANTHEREAL reviewed (commented):
any other problems？ @siddontang

foreverbell committed:
Rename canPrewrite/Commit to canGoPrewrite/Commit.

foreverbell committed:
Also clean up stale secondary locks if primary lock exists.

Added LockConsistency invariant to verify no missing secondary locks
after prewrite phase.

foreverbell commented:
Discussed offline, we should also try to clean up stale secondary locks even if the primary lock still exists. This should be more approximate to the implementation.

Added `LockInvariant` to verify there is no missing secondary locks before commit.

siddontang commented:
should add an abort invariant, if aborted, the primary key can't be committed.

foreverbell committed:
Added AbortedConsistency.

foreverbell commented:
Done.

foreverbell committed:
fix

siddontang commented:
Great work. LGTM

IANTHEREAL commented:
LGTM
